File: ex4/codes/4_5.m
format long
F = 253.86549552;
c = 1.16674;
m = 239.245;
alpha = F / m;
beta = c / m;

xspan = [0 91.44];
[x, V] = ode45(@(x, V) alpha - beta * sqrt(2.0 * V), xspan, 0);
Answer1 = sqrt(2.0 * V(end))

syms v(x);
eqn = diff(v, x) * v == alpha - beta * v;
cond = v(0) == 0;
vSol(x) = dsolve(eqn, cond);
Answer2 = double(vSol(91.44))

syms v;
eqn = alpha * 91.44 + alpha / beta * v + (alpha^2.0) / (beta^2.0) * log(1 - beta / alpha * v) == 0;
Answer3 = double(solve(eqn, v))

File: ex4/codes/4_5.py
from scipy.integrate import odeint
from sympy import symbols, Function, dsolve, solve, log
import numpy as np

F = 253.86549552
c = 1.16674
m = 239.245
alpha = F / m
beta = c / m

xspan = np.linspace(0, 91.44, 100)
V0 = 0


def dV_dx(V, x):
    return alpha - beta * np.sqrt(2.0 * V)


V = odeint(dV_dx, V0, xspan)
Answer1 = np.sqrt(2.0 * V[-1])[0]
print(Answer1)

#! TODO

x = symbols("x")
v = Function("v")(x)
eqn = v.diff(x) * v - alpha + beta * v
cond = {v.subs(x, 0): 0}
vSol = dsolve(eqn, ics=cond)
Answer2 = float(vSol.rhs.subs(x, 91.44))
print(Answer2)

v = symbols("v")
eqn = (
    alpha * 91.44
    + alpha / beta * v
    + (alpha ** 2.0) / (beta ** 2.0) * log(1 - beta / alpha * v)
)
Answer3 = float(solve(eqn, v)[0])
print(Answer3)
